src/ThreadDb.cc

```diff
@@ -0,0 +1,232 @@
+/* -*- Mode: C++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
+
+#include "ThreadDb.h"
+#include "GdbServer.h"
+#include "Task.h"
+#include "TaskGroup.h"
+#include "log.h"
+
+extern "C" {
+// The proc_service/thread_db library has a very weird API.  It
+// requires the user of the library to provide certain functions which
+// it links to (rather than, say, having the library user supply a
+// struct of function pointers).  We have to ensure that these
+// functions have C linkage, so that libthread_db can find them.
+#include "proc_service.h"
+}
+
+#include <asm/prctl.h>
+#include <dlfcn.h>
+#include <linux/elf.h>
+#include <sys/reg.h>
+
+#define LIBRARY_NAME "libthread_db.so.1"
+
+// Needed for the logging API.
+using namespace rr;
+
+ps_err_e ps_pglobal_lookup(struct ps_prochandle* h, const char*,
+                           const char* symbol, psaddr_t* sym_addr) {
+  rr::remote_ptr<void> addr;
+  if (!h->db->query_symbol(symbol, &addr)) {
+    LOG(debug) << "ps_pglobal_lookup " << symbol << " failed";
+    return PS_NOSYM;
+  }
+  *sym_addr = reinterpret_cast<psaddr_t>(addr.as_int());
+  LOG(debug) << "ps_pglobal_lookup " << symbol << " OK";
+  return PS_OK;
+}
+
+ps_err_e ps_pdread(struct ps_prochandle* h, psaddr_t addr, void* buffer,
+                   size_t len) {
+  bool ok = true;
+  uintptr_t uaddr = reinterpret_cast<uintptr_t>(addr);
+  // We need any task associated with the task group.  Here we assume
+  // that all the tasks in the task group share VM.
+  rr::Task* task = *h->task_group->task_set().begin();
+  task->read_bytes_helper(uaddr, len, buffer, &ok);
+  LOG(debug) << "ps_pdread " << ok;
+  return ok ? PS_OK : PS_ERR;
+}
+
+ps_err_e ps_pdwrite(struct ps_prochandle*, psaddr_t, const void*, size_t) {
+  LOG(debug) << "ps_pdwrite";
+  return PS_ERR;
```

Comment:
I think it's probably better to do
````
FATAL() << "ps_pdwrite not implemented";
````
If we hit this, that'll be easier to disagnose than some more mysterious failure.